Added a new section about integrating Prometheus alerts

suyogsainkar opened the pull request:
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1510526

moolitayer commented:
Hi @suyogsainkar

Right now it's only relevant for Containers so I would name the section "Container Prometheus Alerts" or "OpenShift Prometheus Alerts".

Installing Prometheus is outlined in OpenShift documentation:
https://docs.openshift.com/container-platform/3.7/install_config/cluster_metrics.html#openshift-prometheus
Maybe it's possible to link that?

We are missing instructions on how to define alerts, currently has to be done on the Prometheus side:
```
$ oc edit configmap -n openshift-metrics prometheus
# Supported annotations:
# severity: ERROR|WARNING|INFO. defaults to ERROR.
# miqTarget: ContainerNode|ExtManagementSystem, defaults to ContainerNode.
# miqIgnore: "true|false", should ManageIQ pick up this alert, defaults to true.
  prometheus.rules: |
    groups:
    - name: example-rules
      interval: 30s # defaults to global interval
      rules:
      - alert: "Node Down"
        expr: up{job="kubernetes-nodes"} == 0
        annotations:
          miqTarget: "ContainerNode"
          severity: "ERROR"
          url: "https://www.example.com/fixing_instructions"
          message: "Node {{$labels.instance}} is down"
```
Then one must reload the Prometheus configuration, one way of doing this is by deleting the Prometheus pod which would then redeploy with the new configuration.

All of this should be made easier once we go out of tech preview

suyogsainkar commented:
@moolitayer Thanks for your thorough review and comments! I have applied all your suggestions in my latest commits. Would you mind taking a quick look please? Let me know  if you have any other feedback. Thanks!

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,73 @@
+    groups:
+    - name: example-rules
+      interval: 30s # defaults to global interval
+      rules:
```
add here:
```
# copy the alerts below this line
```

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,73 @@
+          miqTarget: "ContainerNode"
+          severity: "ERROR"
+          url: "https://www.example.com/fixing_instructions"
+          message: "Node {{$labels.instance}} is down"
```
We have a another kind of alert (`miqTarget: "ExtManagementSystem"`) that shows up differently so it's probably worth adding an example for that kind as well. To avoid confusion here is the entire block again:
```
# Supported annotations:
# severity: ERROR|WARNING|INFO. defaults to ERROR.
# miqTarget: ContainerNode|ExtManagementSystem, defaults to ContainerNode.
# miqIgnore: "true|false", should ManageIQ pick up this alert, defaults to true.
  prometheus.rules: |
    groups:
    - name: example-rules
      interval: 30s # defaults to global interval
      rules:
      # 
      # ------------- Copy below this line -------------
      #
      - alert: "NodeDown"
        expr: up{job="kubernetes-nodes"} == 0
        annotations:
          miqTarget: "ContainerNode"
          severity: "ERROR"
          url: "https://www.example.com/node_down_fixing_instructions"
          message: "Node {{$labels.instance}} is down"
      - alert: "TooManyRequests"
        expr: rate(authenticated_user_requests[2m]) > 12
        annotations:
          miqTarget: "ExtManagementSystem"
          severity: "ERROR"
          url: "https://www.example.com/too_many_requests_fixing_instructions"
          message: "Too many authenticated requests"
```

suyogsainkar replied:
Ack, updated.

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,73 @@
+Alerts that were triggered before assigning the profiles to the enterprise will not appear in {product-title_short} at all.
+====
++
+. Add a Prometheus alerts endpoint.
```
This is referring to this:
![ssl_error](https://user-images.githubusercontent.com/3010449/34217078-0c5079fc-e5b3-11e7-879f-0b4d2f144621.png)


(there's a bug there that no longer exists, but that's why I tool the screen shot...)

suyogsainkar replied:
Thanks - I believe this includes the procedure to add OpenShift Containers Provider and the Alerts endpoints? I have updated accordingly.

moolitayer commented:
Do we want to include a screenshot or two from the new alerts screen? it looks nice...

suyogsainkar commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+This feature is available with this release of {product-title} as a technology preview only. For more information on the support scope for features marked as technology previews, see https://access.redhat.com/support/offerings/techpreview[Technology Preview Features Support Scope].
+====
+
+Prometheus is used as an external alerting component. {product-title_short} integrates Prometheus alerts for use with OpenShift Container Platform. {product-title_short} collects EmsEvents from Prometheus, generates alerts based on these events, and then attaches alerts to inventory objects.
```
@moolitayer Wanted to check if this intro text correctly captures the current feature and sequence of steps that is useful to the user? Any feedback here?

moolitayer replied:
Looks good. replace `EmsEvents` with `event` I think it's too technical for docs

suyogsainkar replied:
Fixed, thanks.

suyogsainkar commented:
@moolitayer Thanks Mooli! I have updated the section based on your feedback. Also summarised the steps upfront and created subsections for improved readability. What do you think? Also, I do like the idea of adding a screenshot, however we usually avoid adding them unless it is a must. That said, could you share a screenshot or two with everything configured and I'll have a look. I really appreciate all your inputs. Thanks!

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+
+It is possible to view ongoing alerts in {product-title_short} and manage their life cycle, including the ability to:
+
+* View active alerts per provider in the alert dashboard.
```
View active alerts per provider in "Overview" screen.

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+
+Prometheus is used as an external alerting component. {product-title_short} integrates Prometheus alerts for use with OpenShift Container Platform. {product-title_short} collects EmsEvents from Prometheus, generates alerts based on these events, and then attaches alerts to inventory objects.
+
+It is possible to view ongoing alerts in {product-title_short} and manage their life cycle, including the ability to:
```
It is possible to view ongoing alerts in {product-title_short} under the "Monitor" section, and manage their life cycle :

suyogsainkar replied:
Updated, thanks.

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+
+* View active alerts per provider in the alert dashboard.
+* View alert data and related objects.
+* Assign and acknowledge alerts in the alerts list screen.
```
Comment, assign and acknowledge alerts in the "All Alerts" screen.

suyogsainkar replied:
Updated, thanks.

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+
+. Deploying and configuring Prometheus on OpenShift.
+. Assigning Prometheus alert profiles to the enterprise in {product-title_short}.
+. Adding OpenShift Containers Providers and Prometheus alert endpoints in {product-title_short}.
```
:+1:

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+==== Deploying and Configuring Prometheus on OpenShift
+
+. Deploy Prometheus on OpenShift by following the steps covered in link:https://docs.openshift.com/container-platform/3.7/install_config/cluster_metrics.html#openshift-prometheus[Prometheus on OpenShift Container Platform].
+. Add Prometheus to an OpenShift cluster and configure alert definitions. See the example code block below for configuring alerts (currently must be done on the Prometheus side):
```
:+1:

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
++
+
+
+==== Adding OpenShift Containers Providers and Prometheus Alert Endpoints
```
Adding OpenShift Containers Provider with a Prometheus Alert Endpoint

moolitayer replied:
This while section should be under the general endpoints section... and maybe linked from here...

suyogsainkar replied:
sorry, i don't understand. could you clarify this a bit more please. you mean this entire procedure/section should be moved under an existing general endpoints section and no need to have this section Adding an OpenShift Containers Provider with Prometheus Alert Endpoints here? perhaps, update and link section - https://github.com/ManageIQ/manageiq_docs/blob/6ef2bef2d9cfddbceb5e1617ccacae626cb15fe5/doc-Managing_Providers/topics/Editing_a_Containers_Provider.adoc - from here as you've commented below?

moolitayer replied:
Exactly , we already have that general endpoint section. It is a little out of sync though.
You can see there we have:
"5. Under Endpoints in the Default tab, edit the following as required:"
Remained unchanged.

"6. Under Endpoints in the Hawkular tab, configure the following for Hawkular capacity and utilization metrics collection:"

That tab in the next version is actually called "Metrics" and you can choose Hawkular OR Prometheus

There should be another section named
"6. Under Endpoints in the Alerts tab, configure the following for Prometheus alerting from the cluster"
...

Just like the existing sections for the new alert endpoint


Hope that helps.

suyogsainkar replied:
@moolitayer thanks for the clarification, understood. 
In that case, 
1. I'll apply updates/feedback from here (mainly in relation to the steps about Trusted CA Certificate and Token) to section https://access.redhat.com/documentation/en-us/red_hat_cloudforms/4.6-beta/html-single/managing_providers/#adding_openshift_provider in the Managing Provider's guide and link that section from under 9.1.3. Adding an OpenShift Containers Provider with a Prometheus Alert Endpoint as the third part of 9.1. Configuring Prometheus Alerts in CloudForms (instead of linking the Editing a Containers Provider section) which I think would be a better fit as part of the end-to-end steps of Configuring Prometheus Alerts in CF). 
2. Then, per your above comment - I'll also update the section https://access.redhat.com/documentation/en-us/red_hat_cloudforms/4.6-beta/html-single/managing_providers/#editing_a_containers_provider to include the missing step: "7. Under Endpoints in the Alerts tab, configure the following for Prometheus alerting from the cluster"
...  What do you think?
3. Also, I know it's not directly related to this PR/topic, but while I'm at it - wondering if Hawkular is now completely gone and removed from the Metrics dropdown? So that I can update the procedure and remove Hawkular references accordingly.
Thanks, -Suyog

moolitayer replied:
Yes the "Hawkular" tab is completely gone, we now have a "Metrics" tab with two options "Prometheus" and "Hawkular"

cc @yaacov

@suyogsainkar Do have a running 5.9? ping me if you want to take a look at a running system

suyogsainkar replied:
Thanks @moolitayer. Yes, I have access to a running 5.9 (version 5.9.0.9.20171115202245_7429f75), though I'll soon be updating to the latest available build. I can see the Metrics tab with the two options Prometheus and Hawkular. However I wasn't sure if there's any change with the Hawkular option in there - got confused with the removal of the middleware/hawkular provider in CF, which is not related to this. Thanks for confirming.

suyogsainkar replied:
![ocp-add](https://user-images.githubusercontent.com/6856557/34705516-aabdd988-f54c-11e7-82eb-0d366f4b3f3a.png)

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+
+  # oc get secret --namespace management-infra management-admin-token-8ixxs --template='{{index .data "ca.crt"}}' | base64 --decode
+  
+Paste the output (a block of text starting with `-----BEGIN CERTIFICATE-----`) into the *Trusted CA Certificates* field.
```
there is a mixup between the ca certificate which is the main openshift CA available in `/etc/origin/master/ca.crt` and the management-admin service account. 

I don't this this section should exist here only what we have in the endpoints sections, for example: https://github.com/ManageIQ/manageiq_docs/blob/6ef2bef2d9cfddbceb5e1617ccacae626cb15fe5/doc-Managing_Providers/topics/Editing_a_Containers_Provider.adoc
I see that page needs some changes as well... not sure if related to this work but we no longer have a "hawkular tab" but "Metrics" and "Alerts" tabs...

suyogsainkar replied:
"there is a mixup between the ca certificate which is the main openshift CA available in /etc/origin/master/ca.crt and the management-admin service account."
- could you clarify this more please? so which of the two methods to obtain the CA certificate should be covered here?

"I don't this this section should exist here only what we have in the endpoints sections, for example: https://github.com/ManageIQ/manageiq_docs/blob/6ef2bef2d9cfddbceb5e1617ccacae626cb15fe5/doc-Managing_Providers/topics/Editing_a_Containers_Provider.adoc"
-Continuing from the above comment, so you mean the current section is not needed here as the user needs to only edit an existing containers provider to include the step on alert endpoint (alert tab) as part of configuring and enabling Prometheus alerts in CloudForms? so edit provider and add this:
----
Add the Prometheus alerts endpoints in the Alerts tab:

a.Select a Security Protocol method to specify how to authenticate the service:

SSL: Authenticate the provider securely using a trusted Certificate Authority. Select this option if the provider has a valid SSL certificate and it is signed by a trusted Certificate Authority. No further configuration is required for this option.
SSL trusting custom CA: Authenticate the provider with a self-signed certificate. For this option, copy your provider’s CA certificate to the Trusted CA Certificates box in PEM format.
SSL without validation: Authenticate the provider insecurely using SSL. (Not recommended)
b. Enter the Hostname or IPv4 or IPv6 address of the provider.
c. Enter the API Port if your Prometheus provider uses a non-standard port for access. The default port is 443.
d. Click Validate to confirm that CloudForms can connect to the alerts service.
Click Add.
----
I was considering making this as an end-to-end example which includes all of the steps. 

"I see that page needs some changes as well... not sure if related to this work but we no longer have a "hawkular tab" but "Metrics" and "Alerts" tabs..." 
-Ack, will create a separate bug to track / update that section, also since hawkular no longer exists. thanks.

moolitayer replied:
> could you clarify this more please? so which of the two methods to obtain the CA certificate should be covered here?

There are two separate things that two separate places in the UI
"Trusted CA Certificates" is relevant to all endpoints (default, metrics, alerts) they can all be obtained from the `/etc/origin/master/ca.crt`

"Token" Is what you get from `oc get secret` command, thats relevant only for the default endpoint.

suyogsainkar replied:
@moolitayer Thanks for the clarification. I've slightly modified the two steps to avoid any mix up. Hope it makes sense.

moolitayer commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+. Click *Save*.
+. Click to expand *Provider Alert Profiles*, then click *Prometheus Provider Profile*. 
+. Click *Configuration*, then click *Edit assignments for this Alert Profile*. Assign the profile to *The Enterprise* from the *Assign To* list.
+. Click *Save*.
```
:+1:

moolitayer commented:
Looks good, I've added some comments especially about the add endpoint section that should be updated in the existing container endpoints sections and not added in this new document.

cc @shalomnaim1

shalomnaim1 commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+* *SSL*: Authenticate the provider securely using a trusted Certificate Authority. Select this option if the provider has a valid SSL certificate and it is signed by a trusted Certificate Authority. No further configuration is required for this option.
+* *SSL trusting custom CA*: Authenticate the provider with a self-signed certificate. For this option, copy your provider’s CA certificate to the *Trusted CA Certificates* box in PEM format.
+* *SSL without validation*: Authenticate the provider insecurely using SSL. (Not recommended)
+.. Enter the *Hostname* or IPv4 or IPv6 address of the provider. 
```
You have enter here the alert's route from OCP

suyogsainkar replied:
@shalomnaim1 sorry, could you clarify this please?

shalomnaim1 replied:
yes,

To configure alerts endpoint you have to use Prometheus alerts route.

you can get this route by running  ``` oc get routes --all-namespaces``` in your OCP master and copy openshift-metrics alerts URL

```{shell}
[root@snaim-ocp-3-7-master ~]# oc get route  --all-namespaces 
NAMESPACE           NAME               HOST/PORT                                                                  PATH      SERVICES           PORT      TERMINATION   WILDCARD
default             docker-registry    docker-registry-default.snaim-ocp-3-7.cfme2.lab.eng.rdu2.redhat.com                  docker-registry    <all>     passthrough   None
default             registry-console   registry-console-default.snaim-ocp-3-7.cfme2.lab.eng.rdu2.redhat.com                 registry-console   <all>     passthrough   None
openshift-infra     hawkular-metrics   hawkular-metrics.snaim-ocp-3-7.cfme2.lab.eng.rdu2.redhat.com                         hawkular-metrics   <all>     reencrypt     None
openshift-metrics   alerts             alerts-openshift-metrics.snaim-ocp-3-7.cfme2.lab.eng.rdu2.redhat.com                 alerts             <all>     reencrypt     None
openshift-metrics   prometheus         prometheus-openshift-metrics.snaim-ocp-3-7.cfme2.lab.eng.rdu2.redhat.com             prometheus         <all>     reencrypt     None
[root@snaim-ocp-3-7-master ~]# 
```
In this case, this would be ```alerts-openshift-metrics.snaim-ocp-3-7....```

moolitayer replied:
I would not go into such detail.

If they have the route with the name & namespace we expect the `detect` button would work. Otherwise I would just write enter the `Hostname` (like it is in the UI) or `Route`

suyogsainkar replied:
Understood, thanks.

shalomnaim1 commented on doc-Monitoring_Alerts_and_Reporting/topics/Integrating_Prometheus_Alerts.adoc:
```diff
@@ -0,0 +1,130 @@
+    - name: example-rules
+      interval: 30s # defaults to global interval
+      rules:
+      - alert: "NodeDown"
```
Add a "<<" sign to mark the start of the alert section

shalomnaim1 replied:
No that way, I meant to add this like this

```{yaml}
     - name: example-rules
        interval: 30s # defaults to global interval
        rules:
            - alert: "NodeDown" << Alert start here
```

suyogsainkar commented:
I've incorporated all suggestions per tech review discussion. Also updated the steps in the adding and editing container providers section of the Managing Providers guide to address some inaccuracies and the UI changes.

Updated preview: http://file.bne.redhat.com/ssainkar/Prometheus-v2/html-single/#integrating_prometheus_alerts

@cbudz Hey Chris, would you mind peer reviewing and merging this PR if everything looks OK. Let me know if you spot any errors, or if you have any other feedback. I'll cherrypick updates to the 4.6 branch once merged. Thanks! - Suyog

cbudz commented:
Hey Suyog,

This looks great. Merging to master.

-Chris
